### PR TITLE
Add initial documentation for ARC Controller configuration

### LINKED ISSUE #2388: Make etcd and prometheus endpoints optional in configuration
- [x] Etcd endpoint optional in agent configuration – done in https://github.com/fluxninja/aperture/pull/2339
- [ ] Prom endpoint optional in agent configuration
- [ ] Etcd endpoint optional in agent CR
- [ ] Prom endpoint optional in agent CR

### PR DESCRIPTION
**Pull request by kklimonda-fn**
Fixes #2388 

##### Checklist

- [x] Tested in playground or other setup
- [x] Documentation is changed or added

### Summary by CodeRabbit

Release Notes:
- Documentation: Added initial documentation for ARC Controller configuration.
- Documentation: Updated the documentation for ARC Controller configuration by removing the "required" constraint from two fields.
- New Feature: Added a new template function `agent.fluxninja.endpoint` to create the endpoint of the FluxNinja ARC for Aperture Agent. It checks if the `fluxninja.endpoint` value is provided and fails if it's empty.
- Bug fix: Added validation checks for the presence of etcd endpoints and Prometheus address in the ARC Controller configuration. Refactored error handling and added validation checks for certain fields.
- Bug fix: Added a check to ensure that the etcd endpoints are provided in the configuration. Updated the `Endpoints` field in the `EtcdConfig` struct to be optional during JSON serialization.
- Bug fix: Modified the `validate` tag for the `Address` field in the `PrometheusConfig` struct to allow an empty value.
- Bug fix: Added a condition to check if `promClient` is not nil before adding the Prometheus remote write exporter and updating the pipeline configuration.
- Documentation: Update the documentation for ARC Controller configuration, providing instructions for both ARC Controller and Aperture Controller deployments.
- Documentation: Added a documentation link to the `EnableCloudController` field in the `FluxNinjaExtensionConfig` struct.
- Chore: Added a temporary fix for an issue with the `Cython` package version. No other changes were made.
- Chore: Added installation of Python tools and updated PyYAML version.

> "Code changes dance,
> Documentation sings,
> Bugs fixed, harmony rings.
> New features take a chance,
> Celebrating this PR's advance!"

### CODE REVIEW

File: cmd/aperture-agent/agent/otel-config.go [hunk 1]
```diff
@@ -166,8 +166,13 @@ func addMetricsPipeline(
 	lis *listener.Listener,
 	promClient promapi.Client,
 ) {
+	pipeline := otelconfig.Pipeline{}
 	addPrometheusReceiver(config, agentConfig, tlsConfig, lis)
-	otelconfig.AddPrometheusRemoteWriteExporter(config, promClient)
+	if promClient != nil {
+		otelconfig.AddPrometheusRemoteWriteExporter(config, promClient)
+		pipeline.Exporters = []string{otelconsts.ExporterPrometheusRemoteWrite}
+	}
```
Comment (krdln) [inline cmd/aperture-agent/agent/otel-config.go:174]: Can you please check this with custom infra meters? We're using the same exporter `otelconsts.ExporterPrometheusRemoteWrite` there and I wonder if there's no problem with referring to non-existing exporter

File: operator/controllers/agent/reconciler.go [hunk 2]
```diff
@@ -408,37 +408,29 @@ func (r *AgentReconciler) updateAgent(ctx context.Context, instance *agentv1alph
 
 // checkDefaults checks and sets defaults when the Defaulter webhook is not triggered.
 func (r *AgentReconciler) checkDefaults(ctx context.Context, instance *agentv1alpha1.Agent) error {
-	resource, err := r.DynamicClient.Resource(api.GroupVersion.WithResource("agents")).Namespace(instance.GetNamespace()).Get(ctx, instance.GetName(), v1.GetOptions{})
-	if err != nil {
+	updateStatus := func(instance *agentv1alpha1.Agent, reason, message string) error {
 		instance.Status.Resources = controllers.FailedStatus
-		r.Recorder.Eventf(instance, corev1.EventTypeWarning, "FailedToFetch", "Failed to fetch Resource. Error: '%s'", err.Error())
+		r.Recorder.Eventf(instance, corev1.EventTypeWarning, reason, message)
```
Comment (krdln) [inline operator/controllers/agent/reconciler.go:413]: Are these changes related?

File: operator/controllers/agent/reconciler.go [hunk 3]
```diff
@@ -447,24 +439,24 @@ func (r *AgentReconciler) checkDefaults(ctx context.Context, instance *agentv1al
...
+			return updateStatus(instance, "ValidationFailed", "At least one etcd endpoint must be provided under spec.config.etcd.endpoints.")
 		}
-		return nil
+
+		if instance.Spec.ConfigSpec.Prometheus.Address == "" {
+			return updateStatus(instance, "ValidationFailed", "The address for Prometheus must be provided under spec.config.prometheus.address.")
+		}
+	} else if instance.Spec.ConfigSpec.FluxNinja.Endpoint == "" {
+		return updateStatus(instance, "ValidationFailed", "The endpoint for Flux Ninja must be provided under spec.config.fluxNinja.endpoint.")
+	}
```
Comment (krdln) [inline operator/controllers/agent/reconciler.go:452]: Nit: can we somehow reuse the code from hook.go?

File: operator/controllers/controller/reconciler.go [hunk 4]
```diff
@@ -398,57 +398,45 @@ func (r *ControllerReconciler) deleteResources(ctx context.Context, log logr.Log
...
-		return nil
+		return updateStatus(instance, "FailedToSetDefaults", fmt.Sprintf("Failed to set defaults. Error: '%s'", err.Error()))
+	}
+
+	if len(instance.Spec.ConfigSpec.Etcd.Endpoints) == 0 {
+		return updateStatus(instance, "ValidationFailed", "At least one etcd endpoint must be provided under spec.config.etcd.endpoints.")
+	}
+
+	if instance.Spec.ConfigSpec.Prometheus.Address == "" {
+		return updateStatus(instance, "ValidationFailed", "The address for Prometheus must be provided under spec.config.prometheus.address.")
```
Comment (krdln) [inline operator/controllers/controller/reconciler.go:431]: Ditto

File: pkg/prometheus/client.go [hunk 5]
```diff
@@ -53,6 +57,15 @@ type ClientIn struct {
 }
 
 func providePrometheusClient(in ClientIn) (prometheusv1.API, promapi.Client, error) {
+	var fluxNinjaConfig extconfig.FluxNinjaExtensionConfig
+	if err := in.Unmarshaller.UnmarshalKey(fluxninjaCOnfigKey, &fluxNinjaConfig); err != nil {
+		log.Error().Err(err).Msg("unable to deserialize")
```
Comment (krdln) [inline pkg/prometheus/client.go:62]: Nit: Either change to warn and be more descriptive what fails, or just ignore the error altogether

File: operator/controllers/agent/reconciler.go (hunk 3 above)
Comment (hdkshingala) [reply operator/controllers/agent/reconciler.go:452]: The checks are same but the return and all are different there so for now let's keep them as is and we can pick this later

File: operator/controllers/controller/reconciler.go (hunk 4 above)
Comment (hdkshingala) [reply operator/controllers/controller/reconciler.go:431]: The checks are same but the return and all are different there so for now let's keep them as is and we can pick this later

File: cmd/aperture-agent/agent/otel-config.go (hunk 1 above)
Comment (hdkshingala) [reply cmd/aperture-agent/agent/otel-config.go:174]: Made the changes to handle this

File: operator/controllers/agent/reconciler.go (hunk 2 above)
Comment (hdkshingala) [reply operator/controllers/agent/reconciler.go:413]: I made it a common function as the same was used multiple times

File: manifests/charts/aperture-agent/crds/fluxninja.com_agents.yaml [hunk 6]
```diff
@@ -1299,9 +1299,17 @@ spec:
...
                         description: Whether to enable cloud controller. Overrides
                           etcd and TLS configurations. See [FluxNinja ARC](/arc/arc.md)
                           for more details.
+||||||| parent of 4d21a3701 (CI fixes and review comment changes)
+                        description: Whether to enable cloud controller. Overrides
+                          etcd and TLS configurations.
+=======
+                        description: Whether to enable ARC controller. Overrides etcd,
+                          prometheus writer and TLS configurations.
+>>>>>>> 4d21a3701 (CI fixes and review comment changes)
```
Comment (krdln) [inline manifests/charts/aperture-agent/crds/fluxninja.com_agents.yaml:1312]: unresolved merge conflicts :D please regen pre-commit :)

**Review by krdln — APPROVED** [review]

File: pkg/prometheus/client.go [hunk 7]
```diff
@@ -53,6 +57,11 @@ type ClientIn struct {
 }
 
 func providePrometheusClient(in ClientIn) (prometheusv1.API, promapi.Client, error) {
+	var fluxNinjaConfig extconfig.FluxNinjaExtensionConfig
+	if err := in.Unmarshaller.UnmarshalKey(fluxninjaCOnfigKey, &fluxNinjaConfig); err == nil && fluxNinjaConfig.EnableCloudController {
+		return nil, nil, nil
```
[the remark this answers is not in the document]
Comment (krdln) [reply pkg/prometheus/client.go:62]: @hdkshingala Consider adding a comment and/or a log here so it's more clear why we're returning nils

File: manifests/charts/aperture-agent/crds/fluxninja.com_agents.yaml (hunk 6 above)
Comment (hdkshingala) [reply manifests/charts/aperture-agent/crds/fluxninja.com_agents.yaml:1312]: Fixed it

**Pull request merged**